[DO NOT MERGE] [WIP] StringGuts

milseman opened the pull request:
<!-- What's in this pull request? -->

An initial implementation of a more resilient 2-word representation for String. It uses BridgeObject (for now) to achieve a high degree of resilience. In the future, we'll add more small string forms, establish UnsafeString, and generally rejigger String's views to be properly layered on top.

Notes:

* Self slices are currently implemented by forming `_NSContiguousStrings` to hold the sliced up _StringCore. This incurs overhead. Long term, we could support owned self slices, but for now it would be fruitful to try to eliminate all uses of self-slicing _StringCore from the stdlib. The most common occurrence (but by no means the only) is through _ephemeralString. Instead, for anything that can't possibly escape, we should convert to an UnsafeString, guarantee the lifetime, and slice that up.

* We cannot utilize small representations effectively until we start to wean ourselves off of `_LegacyStringCore`. Ideally (depending on project planning needs) we would be fully weaned off before adding small representations. We'd also like to get close to performance parity where it matters prior, to better gauge the impact of such optimizations.

* The spiritual predecessor to this effort is at https://github.com/dabrahams/swift/commits/string-recore, and while the physical implementation differs, that's a great demonstration of what we're trying to accomplish.

milseman committed:
[stdlib] Rename _StringCore to _LegacyStringCore. NFC.

In grand LLVM tradition, the first step to redesigning _StringCore is
to first rename it to _LegacyStringCore. Subsequent commits will
introduce the replacement, and eventually all uses of the old one will
be moved to the new one.

NFC.

milseman committed:
WIP: Needs cleanup before merge: BridgeObject/NativeObject helper functionality

WIP: StringGuts: Prototype initial implementation. Very slow.

WIP: SQUASH ME: DCE

WIP: Carve out the necessary flag bits; less reliance on old _StringCore

WIP: Views store guts rather than cores

WIP: create an UnsafeString from other representations

WIP: sqash me

WIP: Native append working, but not doing uniqueness check

WIP: Implement uniqueness and byte width checks in append

Also remove unusued return value of _formNative(forAppending:).

WIP: Implement width extension in _UnsafeString._copy.

WIP: Switch to _StringGuts.append(_:)

WIP: String.utf16: Create view from guts directly

milseman commented:
Note, some tests are testing _StringCore's ABI, which is obviously changed here. They will be failing until updated.

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please benchmark

dabrahams commented:
It seems like a shame to throw out all this work, which performs well since Arnold's fixes went in: https://github.com/dabrahams/swift/commits/string-recore

milseman commented:
This is inspired by that work (in fact one of the commits explicitly linked to it, but I think that got dropped during a recent squashing). More and more of it will be moved over, as needed. The work can't be taken verbatim, as it's untenable for ABI stability, at least as the Swift platform exists today.

As I'm sure you're well aware, we don't want the project to continuously succumb to the sunk cost fallacy. Not all code lives forever. Do you have any specific, concrete concerns here?

dabrahams commented:
I guess my main concern was the use of BridgeObject instead of an enum, and I confess I don't understand why there would be any special resilience issues here.

milseman commented:
Ah, yes. `BridgeObject` gives us precise control over the bit representation. This is useful for resilience because it affords us a means of relaxing resilience bit-by-bit to an extent that wouldn't be offered through an enum. At least, without significant contortions.

An enum might make sense in the future (perhaps for other types, as we hopefully will be "done" with String at some point). The issue today is that enum layout is not settled, and I need to break a dependency between optimal enum layout and the new String.

While this *is* software, and in software all decidable things are possible, I'm trying to carefully weigh the ugliness of `BridgeObject` over the benefits of being decoupled from parts of the system that are still in flux. `BridgeObject` gives us the following kind of pseudo-structure with a smaller "attack surface" of potential breakage:

```swift
enum _StringGuts {
  case ManagedNative(UInt6, NativeObjectAddressBits, UInt3, UInt64)
  case ManagedCocoa(UInt6, ForeignObjectAddressBits, UInt3, UInt64)
  case Value(UInt127)
}
```

Of course, `UInt6`, `UInt127`, `NativeObjectAddressBits`, `ForeignObject`, `ForeignObjectAddressBits`, `UInt3`, `UInt127` all don't exist. Enum layout as it currently exists fails to find and use spare bits as it needs to. Even though this enum has many extra inhabitants, enum layout as it currently exists won't use them for Optional. Any progress made on enum layout in Swift 4.x or Swift 5 should be focused on those enums that must be ABI stable in that timeframe (highest priority), while optimal enum layout for very complicated cases like this can come later, worst case.

milseman commented:
@swift-ci please benchmark

milseman commented:
Argh.

milseman commented:
@swift-ci please smoke benchmark

dabrahams commented:
What are “self“ slices, and why are they interesting?  Is there a reason a wrapper over RangeReplaceableMutableBidirectionalSlice<String> is insufficient as the slice of a String?

lorentey committed:
_StringBuffer: Ensure capacity isn’t less than requested in the ASCII case

In the ASCII case, the newly allocated buffer’s capacity was set to
the actual allocated capacity - 2, which may be 1 less than what was requested.

Fix this by removing capacityBump. The actual capacity will be the requested capacity rounded up to the nearest even number (plus any allocation overhead).

It looks like capacityBump was intended to guarantee that there’s always an extra hidden byte at the end of any ASCII buffer; however, that seems to be unnecessary now. (This is unrelated to nul termination.)

lorentey committed:
UnsafeString: Implement narrowing copy.

lorentey committed:
Fix some byteWidth inconsistencies.

- Make sure _StringGuts.byteWidth matches _StringBuffer.elementWidth in more cases
- Set _StringBuffer.usedEnd correctly in the two-byte case
- Reserve an extra code unit for storing a terminating NUL character in the future
- Use the correct byte width when making a copy before append

lorentey committed:
[DO NOT MERGE]: String._dump() and StringGuts.checkInvariants()

lorentey committed:
Tighten capacity checks.

Don’t reserve extra space for a nul terminator yet; we can
always add code for it later.

lorentey committed:
Fix OpaqueCocoaString; implement opaque code paths.

lorentey committed:
Fix isUniqueNative()

lorentey committed:
Character: Use guts when initializing from a large grapheme cluster

lorentey committed:
unmanged => unmanaged

lorentey committed:
Allow single-byte non-tagged Cocoa strings & cache base address in _StringGlue._otherBits.

TODO: Double check that bit 61 is actually available in the Cocoa ABI
for non-tagged object references

lorentey committed:
UnsafeString: Add subscripts, _pointer(toElementAt:)

lorentey committed:
_memcpy: Make src parameter a non-mutable pointer.

lorentey committed:
UnsafeString: Switch baseAddress to an UnsafeRawPointer

For the most part, UnsafeStrings should be considered read-only.

For the few cases when it is known to be safe to mutate them,
we can simply call UnsafeMutableRawPointer.init(mutating:).
This makes it clear that we’re doing something unusual.

lorentey committed:
Switch _NSContiguousString to use _StringGuts as its storage.

This eliminates _LegacyStringCore from String bridging operations.

lorentey committed:
Migrate String.hasPrefix, String.hasSuffix to _StringGuts.

lorentey committed:
String: Move stored properties to top

lorentey committed:
_StringGuts-based empty initializer, builtin literal init, append operators

lorentey committed:
_StringGuts: Add subscripting access to UTF16 characters

lorentey committed:
Add _StringGuts.reserveCapacity.

lorentey committed:
UnsafeString.unsafeOneByteString => .asciiBuffer, .unsafeUTF16String => .utf16Buffer

lorentey committed:
Remove _cocoaStringToSwiftString_NonASCII(_:)

_makeCocoaStringGuts(_:) should have very similar performance, and it
also does the right thing for tagged objects.

lorentey committed:
Rename internal Cocoa API helpers

Add customary underscore; remove parameter label.

getCocoaPointer(_cocoaString:) => _getCocoaStringPointer(_:)
makeCocoaStringGuts(_cocoaString:) => _makeCocoaStringGuts(_:)

lorentey committed:
Fix some warnings.

lorentey committed:
Migrate CharacterView to _StringGuts (WIP)

lorentey committed:
NativeString: Use _StringBuffer as storage.

This may make things a little easier to optimize.

milseman commented:
@dabrahams, I lost my train of thought and why I was being careful to not reject them out-of-hand. I don't think owned self slices are very interesting and would prefer use of one of the slice types. Unowned self slices are far more useful and interesting: a lexically scoped analogue of `ephemeralString`, forming Strings around immortal storage, etc. Such a construct can also enable optimizations in conjunction with more ownership annotations/tracking in the language's future evolution. What I'm mostly referring to in this PR is an artifact of interop with the old `_StringCore`, whose representation supports them.

But, this is software and anything's possible with enough shenanigans. The argument for owned self slices would be so that String-as-common-currency-type can represent a Substring. This is probably a bad idea, for reasons debated in the one-type vs two-type arguments last year.

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
[stdlib] Use StringGuts directly for isASCII checks

WIP SQUAHS ME: Convert precondition to sanity checks

UnsafeString is not API, and even if it were we probably want
_sanityCheck, or at the very least _debugPrecondition rather than
_precondition.

WIP: Some initial substring equality checking, still needs more work on String

lorentey committed:
Move bit tests into targeted properties

This makes it easier to generate good code for the _unmanagedContiguous property.

milseman committed:
[stdlib] Make sure to store the immutable CFString copy when bridging

WIP: Switch more internals over to _StringGuts directly for now

dabrahams commented:
I'm afraid I don't know how to read that because you didn't answer my first question: What are “self“ slices?

milseman committed:
String Guts: Implement String/Substring equality directly

WIP: don't form a non-tagged cocoa string from a tagged cocoa string

milseman commented:
Aha, that is a confusing term. What I mean by "self slice" is a `String` that's really a slice. That is, its `_StringCore`, which is capable of being its own slice type, is being used as its own slice type.

milseman committed:
StringGuts: Eliminate _LegacyStringCore ping-ponging in Substring(_:String)

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
_StringGuts: Add direct initializer from _StringBuffer.

Going through NativeString could introduce unnecessary ARC traffic.

lorentey committed:
_StringGuts: Refactor init from _LegacyStringCore

- Extract the opaque Cocoa case to a distinct path
- Use _copyToStringBuffer to make copies of self-slices in all cases
- Eliminate _isNativeSelfSlice; comparing counts is sufficient

lorentey committed:
_StringGuts: Reduce Optional wrapping

This potentially reduces ARC traffic.

lorentey committed:
_StringGuts: Remove hasCocoaBuffer flag.

We don’t actually use it anywhere, and provenance of e.g. string literals is ambivalent anyway. (This latter issue caused problems in _sanityCheckIdentical.)

lorentey committed:
_StringGuts: Fix capacity growth for native strings

Previously the old capacity wasn’t taken into account when the string was not
uniquely referenced or when it did not match the expected storage width.

milseman committed:
StringGuts: Port over < and hashValue.

lorentey committed:
_StringGuts: Fix isUniqueNative()

Apparently the order of clauses matters in certain contexts.

lorentey committed:
Remove stray debugging dumps

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
Fix String’s += operator not to discard reserved capacity

Previously += would replace storage of the left operand when it was empty, as an optimization. Restrict this optimization to cases where the left operand has no extra capacity preallocated; also make sure it is also applied in String.append(_:).

rdar://35589243

lorentey committed:
StringProtocol._unmanagedContiguous: Don’t blow up for illegal conformances

lorentey committed:
_StringGuts: Make _isOpaque inlineable for now

milseman committed:
StringGuts: Speed up _unmanagedContiguous by ~30%.

_unmanagedContiguous is starting to be a real perf bottle neck, which
is entirely unnecessary. These changes improve codegen resulting in
about a 30% speedup, while the rest is due to excessive ARC traffic
(future work).

milseman committed:
WIP: Use more guts and fewer cores

WIP: Micro-optimize _unmanagedContiguous a little more

WIP: use more guts and fewer cores

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
StringProtocol, _SwiftStringView: Add _wholeString, _encodedOffsetRange requirements

lorentey committed:
Add _cocoaStringCopyCharacters(from:,range:,into:)

lorentey committed:
joined(separator:): Eliminate usages of _ephemeralStrings

lorentey committed:
_StringGuts: Add support for copying/appending slices of a string

lorentey committed:
Improve fast path of String.append(contentsOf:)

- Switch from _ephemeralContent to _wholeString/_encodedOffsetRange.
- Don’t discard previously reserved storage.

lorentey committed:
Character: Add typealias for the small inline storage type.

lorentey committed:
_StringGuts: Add support for appending characters and code units

lorentey committed:
_NSContiguousString: Cosmetics

lorentey committed:
_NSContiguousString: Add missing precondition.

lorentey committed:
String: Add public overloads on append(contentsOf:).

This should speed things up if people accidentally use append(contentsOf:) with a String or Substring argument.

milseman committed:
StringGuts: UnsafeString is a RandomAccessCollection

milseman committed:
StringGuts: Implement subscript directly, speed up Character fetching

milseman committed:
WIP: Implement sad, sad optimization.

Adding extra branching and @inline(never) function intermediaries
speeds up String index advancing. This is likely because the resultant
code is more easily analyzed by the ARC optimizer, as there need be no
real ARC traffic.

The real speedups will hopefully come by natively implementing these
functions, but this at least peels off one layer of needless ARC.

SQUASH ME: do precondition checking correctly

lorentey committed:
Better fix for _isUniqueNative

Explanation kindly provided by @gottesmm

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
WIP: StringGuts: outline cold code for speedups

milseman committed:
StringGuts: Pull over grapheme-breaking fast-paths on UnsafeString

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
[stdlib][test] NewString: fix some mutability warnings

lorentey committed:
[test] NewString: Substring -> String conversions are expected to make a copy now

lorentey committed:
[test] Update NewString for even newer strings

lorentey committed:
_StringGuts.append(contentsOf:): Update self when done.

This resolves a broken invariant.

lorentey committed:
StringProtocol: Provide default implementations for underscored requirements.

lorentey committed:
[test] Update layout tests for new 16-byte String representation

Both String and Optional<String> are now expected to take 16 bytes on all platforms.

Previously, String was 24 bytes on 64-bit platforms and 12 bytes on 32-bit platforms. Optional<String> was a byte longer on both.

TODO: _StringGuts doesn’t support 32-bit platforms yet.

lorentey committed:
[test] Foundation_NSString.swift: Don’t use internals

lorentey committed:
String._joined: Don’t reserve negative space for empty collections

lorentey committed:
NativeString: Extract storage reallocation logic

This also fixes an issue where NativeString.append() did not ensure storage has the correct byte width to store appended data.

milseman committed:
WIP: Messy: Drop much of _StringGuts.append ARC overhead.

Still more to eliminate

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
StringGuts: bitwise equality implies String equality

milseman committed:
StringGuts WIP: Use fewer cores and more guts

milseman committed:
WIP SQUASH ME: Make trivial bit tests (which are ABI anyways) inlineable

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
UnsafeString: Change SubSequence to RandomAccessSlice.

Unfortunately Collection requires slices to preserve the indices of the original collection, so we can’t use UnsafeString itself as the SubSequence type.

Switch to the default SubSequence (RandomAccessSlice) and add the slice(_:) method as an easy way to generate self-slices.

lorentey committed:
_StringGuts._ensureUniqueNative(increasingCountBy:,minimumByteWidth:): Fix invariant check

lorentey committed:
Fix some warnings.

milseman committed:
StringGuts: Expose more bits, and more direct initializers for StaticString purposes

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
OpaqueCocoaString: Add self-slicing and implement RandomAccessCollection

lorentey committed:
OpaqueCocoaString: Add buffering iterator

lorentey committed:
String.UTF8View: Fix self-slicing semantics in Swift 3 mode

When creating a self-sliced UTF8View, keep track of whether the boundaries of the sliced view fell on character boundaries in the original string. Use this information to emulate the old failable UTF8View -> String initializer that used to return nil in such cases.

lorentey committed:
[test] Make validation-test/stdlib/String.swift compile

Unfortunately the tests looking at reallocation behavior still fail, as the semantics of Swift 3 self-slicing have changed. (We should run the test in Swift 4 mode to test the new behavior.)

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
WIP: Super ugly, needs cleanup: Work around for ARC around _unmanagedContiguous

milseman committed:
WIP: Small tweaking of String.==

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
The above was checking if, in addition to storing the count at a known place, we also store the pointer, making things like `_unmanagedContiguous` near-trivial.

lorentey committed:
Add new String storage class and replace read-only views.

_SwiftStringStorage<CodeUnit> is a new generic class with tail-allocated storage; it replaces _StringBuffer and NativeString. It includes methods for mutating the storage inline, but it isn’t designed for read-only access — use _UnmanagedString for that.

_UnmanagedString<CodeUnit> is a new generic struct providing type-safe read-only view into a contiguous string; it replaces UnsafeString.

_UnmanagedOpaqueString is a new struct replacing OpaqueCocoaString.

The transient structs FatalErrorString, SmallCocoaString and NonTaggedCocoaString  are replaced by _StringGuts properties providing direct access to their fields.

lorentey committed:
_NSContiguousString, _SwiftStringStorage: Implement NSString shadow protocol

Having the shadow protocol is a useful failsafe: conforming to it detected that _NSContiguousString failed to correctly implement NSCopying. It also caught a test case that was calling Cocoa API on AnyObject with Swift 2 names.

lorentey committed:
_NSContiguousString: Disable sub-slicing overrides

One way to revive them would be to cache an _UnmanagedString instance directly in an ivar.

lorentey committed:
_StringGuts: Add missing @_inlineable attributes

This greatly improves generated code

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
Eliminate cases of ‘defer { _fixLifetime(a); _fixLifetime(b) }’

lorentey committed:
Add some missing @_inlineable attributes.

lorentey committed:
Fix miscellaneous issues
- Add extra _sanityChecks
- _StringGuts: Simplify _unmanagedRawStart. Remove _unmanagedRawBufferView.
- Character._count: Nano-optimize

lorentey committed:
Review and sanitize recent uses of _fastPath/_slowPath

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
WIP: Work around a generalization failure with explicit @_specialize

milseman commented on stdlib/public/core/StringGuts.swift:
```diff
@@ -0,0 +1,1571 @@
+  var _nativeRawStorage: _SwiftRawStringStorage {
+    @inline(__always) get {
+      _sanityCheck(_isNative)
+      return _bridgeObject(toNative: _object) as! _SwiftRawStringStorage
```
@lorentey, do you need to unflag it (in case it's UTF16)?

milseman committed:
[stdlib] Migrate UTF8View.Iterator to _StringGuts

milseman commented on test/stdlib/NewString.swift:
```diff
@@ -14,45 +14,33 @@ func hexAddrVal<T>(_ x: T) -> String {
+    return "Cocoa("
+      + "owner: \(hexAddrVal(x._owner)), "
+      + "count: \(x.count))"
+  } else if x._isTaggedCocoa {
```
@lorentey I believe the structure changed around here (i.e. there is no more _isTaggedCocoa var).

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
Err, last smoke benchmark failed with
```
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/swift/lib/IRGen/IRGenDebugInfo.cpp:1981:47: error: no member named 'FindAllocaDbgDeclare' in namespace 'llvm'
  if (isa<llvm::AllocaInst>(Storage) && llvm::FindAllocaDbgDeclare(Storage))
                                        ~~~~~~^
1 error generated.
```

I think that might be us lagging behind master too much. I have a branch that's rebased from master (at least, master as of Friday), so I'll see if that works if this one doesn't.

milseman commented:
That was a rebase from Friday. I'll need to do another one. Hopefully that's enough to get the benchmarks running, though.

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
SQUASH ME: use the cleaned up stdlib symbols

milseman committed:
SQUASH ME: merge artifacts

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
Work around excess ARC; store ASCII pointer and count

milseman commented:
@shahmishal is something wrong with the smoke benchmark bots, or is the build failing? Looking at the console I don't see any links to a failed build

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
Update reflection tests for new String and Character internals.

lorentey committed:
_StringGuts: Make more members public for testing

lorentey committed:
_StringGuts: Add append overload for appending a full _StringGuts value

lorentey committed:
_StringGuts: Fix overlapping moveAssign issue in replaceSubrange

lorentey committed:
[test] Migrate glass box tests from _LegacyStringCore to _StringGuts.

lorentey committed:
_StringGuts.replaceSubrange: Fix cached count after in-place modification

lorentey committed:
StringUTF8.makeIterator(): Optimize ASCII count

lorentey committed:
_StringGuts: Make storage width-related properties @_inlineable.

lorentey committed:
_StringGuts.replaceSubrange: Don’t unconditionally widen storage to UTF-16.

lorentey committed:
[test] stdlib/String.swift: Update remaining broken tests

lorentey committed:
[test] The BridgeObject in _StringGuts always counts as a reference

lorentey committed:
Reduce direct use of UTF-16 magic constants; simplify surrogate tests.

lorentey committed:
String.CharacterView: Remove _core.

lorentey committed:
_StringGuts: Slighty relax _isEmptyLiteral

The empty string literal doesn’t use the empty singleton yet.

This change enables String.append's empty shortcut to work in more cases.

lorentey committed:
_StringGuts._extractSlice: Quickly return an empty string when the range is empty

There is little point to allocating String storage with capacity 0.

lorentey committed:
Add basic Unicode scalar operations & iterators to _StringGuts variants. Use them to de-core String.UnicodeScalarView.

lorentey committed:
_UnmanagedOpaqueString: Fix indexing (d'oh)

milseman committed:
[test] Now that String's smaller, adjust corner-case tests.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
[StringGuts] Drop an underflow check in compareASCII.

Relatively minor, but drop the underflow check for calculating count
difference.

milseman committed:
[StringGuts] Drop dynamic type check from _unmanagedXViews

Relatively minor, but gets inlined everywhere. Drop a dynamic type
check, as the type is already encoded in our representation.

milseman committed:
WIP: Scrap ranges in comparision.

This results in about a 30-50% speedup on StringComparison_ascii as of
right now, even before removing the excess ARC overhead.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented on stdlib/public/core/StringGuts.swift:
```diff
@@ -0,0 +1,1653 @@
+  @_inlineable
+  @inline(__always)
+  public // @testable
+  mutating func isUniqueNative() -> Bool {
```
@lorentey. BTW, is this safe and sufficient in a multi-threaded environment? That is, do we instead need to ensure uniqueness for the entire execution of a method, or is that guaranteed / implied by the fact that self itself is local, and any copies would invariably bump the count? I don't know the rules here. CC @aschwaighofer @atrick @moiseev

aschwaighofer replied:
The standard library does not need to guard against language level races. A language level race is just that and needs to be addressed by the user of the standard library.

For example: 

‘’’
var a: String  // global

Thread 1:

a.mutate()

Thread 2:

a.mutate() // race on ‘a’
‘’’

Will race without any additional synchronization in the user program.

lorentey committed:
Express all CharacterView operations in terms of an underlying base String.

lorentey committed:
_FixedArray: Add withUnsafe[Mutable]BufferPointer(_:) methods.

lorentey committed:
_UnmanagedString: Remove obsolete iterator operations

lorentey committed:
Migrate String away from CharacterView to the new grapheme breaking methods in _StringVariant.

lorentey committed:
_StringVariant: Add methods for forwards/backwards grapheme breaking

measureFirstExtendedGraphemeCluster() returns the length of the first extended grapheme cluster in a _StringVariant instance, while measureLastExtendedGraphemeCluster() returns the length of the last.

_StringVariants are self-sliceable in O(1), so these two methods are enough to measure grapheme clusters at any position in a string and to implement forwards/backwards iteration.

lorentey committed:
Add _StringVariant protocol describing common operations on all String storage views

lorentey committed:
_StringGuts: Add more convenient reserveCapacity variant. Fix usages.

lorentey committed:
String.lowercased(), .uppercased(): Decore and enguttenify

lorentey committed:
Fix some 'variable never mutated' warnings.

lorentey committed:
String.withCString(encodedAs:,_:): Migrate to guts

lorentey committed:
String._encode(_:, into:): Migrate to guts.

lorentey committed:
String.init(_storage: _StringBuffer): Don't use _LegacyStringCore.

lorentey committed:
String.init(repeating:,count:): Migrate to new-style storage.

lorentey committed:
Refactor unicode scalar operations; add them to _StringVariant.

lorentey committed:
Migrate remaining uses of _ephemeralString to _wholeString/_encodedOffsetRange

lorentey committed:
Remove String.init(_fixmeLegacyCore:) 🎉

lorentey committed:
_Stdout.write(_:): Migrate to guts

lorentey committed:
String.UTF16View._encodedOffsetRange: Take _offset/_length into account.

milseman committed:
_StringGuts: Introduce _StringObject encapsulation.

Add a _StringObject struct to wrap the BridgeObject and encapsulate
all the low level, bit twiddly, and unsafe operations. This is meant
to be a "transparent" structure, that is all methods and properties
should always be inlined and directly emitted. It just encapsulates
operations that are otherwise tricky to implement correctly.

milseman committed:
StringGuts: Use _StringObject as backing object storage.

Switch all of StringGuts over to StringObject. Simplifies some of
StringGuts but more importantly, this cordons off many tricky and
unsafe operations from the rest of StringGuts.

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@lorentey I want to see how that run goes. I have just rebased from master again, and after I get those results (and assuming my tests pass) I'll push that rebase.

milseman commented on stdlib/public/core/StringGuts.swift:
```diff
@@ -0,0 +1,1355 @@
+    defer { _fixLifetime(self) }
+    var bitPattern = _object.payloadBits
+    return _isUnique_native(&bitPattern)
+  }
```
CC @atrick, this is what we were discussing earlier.

atrick replied:
I have a better idea now how SILGen could be fixed to allow isUnique(&self.property). I'm not sure when someone would get around to fixing it, but filing a bug for any case you come across would be reasonable and actionable.

milseman committed:
StringGuts: Avoid a checked downcast.

Use an unchecked downcast (i.e. sanity check) rather than a
debug-checked one.

milseman committed:
Work around ARC in getting Cocoa string length.

The StringComparison_ascii benchmark has 6-8 unnecessary pairs of
retain/release per loop iteration. This removes 2 pairs.

milseman committed:
StringGuts: Work around more ARC

StringComparison_ascii benchmark has 4-6 unnecessary retain/release
pairs. This removes 2.

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
Update white box tests for new _StringGuts interface

lorentey committed:
[test] Add basic opaque string iterator test

lorentey committed:
Grapheme Breaking: Fix caching of character strides in String.Index

- Don't cache a stride of zero at the endIndex.
- Don't trap on or truncate strides that are greater than 65535 code units.

milseman committed:
StringGuts: dedicated vectorizable memcpy_zext/trunc

milseman commented:
That had a 2x+ impact on StringBuilderLong, which should now magnify the constant-costs we're facing.

@lorentey, I noticed that append ends up loading type metadata even if specialized. My guess is it's the bit width checks inside UnmanagedString._copy. Have you seen issues here?

Also, the String.swift validation test just started failing for me. It's expecting String range-based subscript to do bounds checking, and currently it doesn't. I don't see how this has changed recently, so it's a little confusing for me.

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
SQUASH ME: @_specialize to force specialization.

Specialize grapheme breaking. 2x speedup on StringMatch.

milseman committed:
NSContiguousString: make methods inlineable, skip overflow check

milseman committed:
StringGuts: Annotate ARC hacks with effects(readonly).

When hacking around ARC by using opaque function calls on unsafe
bitpatterns, self still has to be guaranteed in case of arbitrary
effects. This can limit the degree to which the optimizer is allowed
to optimize. Add effects(readonly). As a result, this speeds up (and
even completely trivializes) many benchmarks.

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
[benchmark] DriverUtils: Use String.withCString instead of internal interfaces

lorentey committed:
[test] Migrate remaining uses of _LegacyStringCore to _StringGuts

lorentey commented:
> @lorentey, I noticed that append ends up loading type metadata even if specialized. My guess is it's the bit width checks inside UnmanagedString._copy. Have you seen issues here?
>
> Edit: Actually, what are your plans for UnmanagedString being generic? I'm seeing all kinds of witness table accesses for UInt8 in UmanagedString's measureExtendedGraphemeClusterForward for StringMatch. They may not contribute much in total, but it's probably going to limit the optimizer in other areas as well.

I really want to split `_UnmanagedString` into two non-generic types; the `CodeUnit` tests are not only slow sometimes, but they also look clumsy. `_StringVariant` will reduce code duplication, but we can also add a new internal protocol for operations that are common between the two contiguous views. (There is definitely no need to do any gybbing.)

> Also, the String.swift validation test just started failing for me. It's expecting String range-based subscript to do bounds checking, and currently it doesn't. I don't see how this has changed recently, so it's a little confusing for me.

I'm not quite sure why the test would fail only now, but some of the bounds checking was lost during the migration to unmanaged string views. My new PR restores most of them, but we should also do an audit pass for any checks that are still missing (especially on API that isn't currently included in the test suite). For now, I added methods to `_StringVariant` for explicit boundary checking; I think a better option would be to make bounds checking the default and explicitly opt-in to unchecked access via `foo[unchecked: i]` subscripts -- this can be done as part of the audit.

lorentey committed:
_ValidUTF8Buffer: Fix warnings; implement isEmpty directly.

lorentey committed:
FixedArray: Implement append & working count

lorentey committed:
_StringGuts, _StringVariant: Add unicodeScalar(startingAt:), unicodeScalar(endingAt:)

lorentey committed:
_StringGuts: Implement Sequence with a buffering iterator

Having this simplifies migrating the UTF-8 view away from _LegacyStringCore, but we should probably get rid of it at some point.

Using a 16-element buffer here is probably overkill.

lorentey committed:
String.UTF8View: Migrate to guts

lorentey committed:
Remove _LegacyStringCore 🎉🎉🎉

lorentey committed:
Remove _StringBuffer 🎉🎉🎉

lorentey committed:
_StringGuts.reserveCapacity: Fix capacity calculation on UTF-16 branch

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
StringGuts: Apply ARC hacks to Substring as well.

In comparisons with Substrings, also do the ARC hacks we did with
String.

milseman committed:
StringGuts: Annotate decls with effects(readonly).

Annotating the unmanaged*Views with @effects(readonly) allows
non-inlining callers such as NSContiguousString to avoid having to
conservatively keep self alive for fear of side effects.

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
SQUASH ME: Unfortunate @_specialize addition.

_appendUTF16 is perhaps too general, and only getting partially
specialized based on CodeUnit. This compounds with
Character._SmallUTF16 not getting specialized produces a ton of silly
witness table dispatches.

This shaves off around 30% from StringEdits.

milseman commented:
@swift-ci please smoke benchmark

lorentey committed:
Support for 32-bit platforms (with `AnyObject?` for now) (#15)

* Restore support for 32-bit platforms

On 32-bit platforms, add an extra 32-bit word to both _StringObject and _StringGuts for padding and metadata bit storage.

Padding _StringObject rather than adding two extra words to _StringGuts makes for unattractively discontiguous inline storage. However, _StringObject needs 2 more metadata bits than are available in a 32-bit BridgeObject; shoving an extra word in there was easier than a full code reorganization.

To make use of all this luxurious extra space, on 32-bit platforms, _StringGuts._otherBits always contains the count for contiguous strings, while the new _StringGuts._extraBits stored property contains the start address. This makes extracting an unmanaged view a trivial operation.

* [runtime] SwiftObject, SwiftValue: Simplify -description implementation

Previously, these classes implemented -description and -debugDescription by
1) calling _getSummary in the stdlib to get the Swift String description corresponding to the value, and then
2) calling _convertStringToNSString in the Foundation overlay to convert the resulting String into an NSString.

To hold the Swift String value between the two calls, SwiftObject defined a dummy C++ struct whose layout is supposed match that of Swift Strings.

Instead of updating this struct to the new String representation, this commit collapses the two stages above into a single call into a new generic function, _getDescription (defined in the Foundation overlay), which gets rid of this maintenance chore.

* [runtime] Reflection: Update String representation to follow stdlib changes

* _StringObject: Replace _BridgeObject with AnyObject? on 32-bit platforms

Also, don't store any tagged value in _StringObject on 32-bit -- with AnyObject taking up a full word now, we only have 28 bits available (30 tops), which is not enough to store either a count nor a start address.

* LibcShims: Adjust _swift_stdlib_print_hex for 32-bit platforms.

* [test] Update 32-bit reflection tests

Previous (speculative) update missed a couple of offset/size changes.

* [test] IRGen/lazy_multi_file.swift: Allow for the possibility of an sret param before self.

String is 4 words long on 32 bit platforms, which gets returned via sret on i386.

* [test] DebugInfo/self.swift: Don't expect self to be alloca'd first

On i386, space for a String gets allocated before the self variable.

* _StringGuts: Add _RawBitPattern typealias

* _StringGuts: Fix an overlapping access warning

* _StringVariant: Remove IndexDistance condition

* [runtime] Update mangled symbol

milseman committed:
WIP: fix up test that had conflict in it

milseman commented:
Rebased

@swift-ci please smoke test

milseman commented:
Scratch that, let's see how bad it is:

@swift-ci please test

milseman committed:
Hack around test failure: Will file a JIRA after landing on master

milseman committed:
WIP: Fix test case to accomodate temporarily-bigger i386 String

milseman committed:
WIP: 32-bit string temporarily has an AnyObject? in it, so reflect that in refernce checks

milseman committed:
WIP: Update test for temporary string-by-address on i386 status

milseman committed:
[tests] Tighten up DebugInfo test with more labels/dags

milseman committed:
WIP: SQUASH ME: made mistake in Filecheck regex

milseman committed:
Revert "[tests] Tighten up DebugInfo test with more labels/dags"

This reverts commit c11ec78a65c31904396b311696bfb4fab8a9615f.

milseman committed:
WIP: disable letstring.swift debuginfo test on 32-bit

lorentey committed:
Add support for compiling without the Objective-C runtime. (#17)

* Restore Linux support

On Linux, Cocoa strings aren't supported; however, we still support opaque strings via the new _OpaqueString protocol, whose requirements approximate a minimal NSString API. (One use of this would be to allow NSString from the open source Foundation lib to be bridged directly. This isn't implemented yet.)

_OpaqueString's interface is not a subset of NSString; in particular, the method getCharacters(_:, range:) is replaced with copyCodeUnits(from:,into:), to get around dealing with _SwiftNSRange.

_OpaqueString is not yet supported on Apple platforms.

* _NSStringCore: Fix length requirement

Be consistent with Foundation; NSString.length is imported/defined as `var length: Int`, not `func length() -> UInt`.

* _StringGuts: Add file header

* _StringGuts: Add _objectIdentifier property, for testing allocation behavior

Also, update tests that used to rely on _underlyingCocoaString -- it is only available with the Objective-C runtime.

* Fix lowercased()/uppercased() on Linux.

* [test] Relax check for realloc behavior on widening String storage

Widening a String from single-byte to two-byte storage may increase available capacity.

* Add deprecated compatibility definitions for SPI used by swift-corelibs-foundation

milseman commented:
@swift-ci please test

milseman committed:
WIP: Remove internalDumpHex

milseman committed:
WIP: Drop _BuiltinBridgeObject/Native aliases (was used for prototyping)

milseman committed:
WIP: Formatting, NFC

lorentey committed:
Custom enum for `_StringObject` on 32-bit platforms (#18)

* [32bit] Use a custom enum type instead of AnyObject? in _StringObject

This allows us to eliminate the _StringGuts._extraBits stored property, reducing the size of String to three words on 32-bit platforms.

In exchange, we don't have uniform flag bits on 32-bit platforms any more, likely reducing performance of our existing _StringObject accessors. (The flag bits are now encoded in the enum representation in the tagged cases.)

* [test] Update tests for 3-word Strings on 32-bit platforms

* [test] Work around a spurious test failure in 32-bit builds

* isEmptyLiteral -> isEmptySingleton

* _StringObject: Add sanityChecks to referenceBits and payloadBits

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
[string] Remove @_inlineable from base class init

milseman committed:
WIP: merge in dump helper

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
WIP: Drop unused stub

milseman committed:
SQUASH ME: Adjust comments

milseman committed:
WIP: Fix linux dump facility

milseman commented:
@swift-ci please test Linux Platform

milseman commented:
(MacOS tests still chugging away at https://ci.swift.org/job/swift-PR-osx/2462/console)

milseman commented:
Failures:
```
    Swift(iphonesimulator-i386) :: stdlib/Unicode.swift.gyb
    Swift(iphonesimulator-i386) :: stdlib/TestUUID.swift
```

Weird, when I run locally I don't get those. @lorentey, if you get a chance, could you investigate?

milseman commented:
@swift-ci please test linux platform

milseman commented:
I don't understand how Linux is failing with:
```
error: initializer for class '_SwiftRawStringStorage' is '@_inlineable' and must delegate to another initializer
```
Even after https://github.com/apple/swift/pull/12442/commits/f932bef39670fb2436e2d0de818582a3031cb941. MacOS builds just fine, there's just 2 test failures currently.

milseman commented:
@swift-ci please test os x platform

milseman committed:
WIP: Oops, Linux too

milseman commented:
@swift-ci please test

milseman commented:
(still trying to reproduce the iphonesimulator failure, which works locally for me, I'm trying the bot's exact configuration now)

milseman commented:
@lorentey, the only remaining failure is on Linux, it seems that a bounds check is not happening: `ForeignIndexes/subscript(Range)/OutOfBoundsTrap/1`. Were you able to get a full validation run on Linux clean, and do you recall if there was any issue here?

https://ci.swift.org/job/swift-PR-Linux/2490/testReport/junit/Swift(linux-x86_64)/stdlib/String_swift/

lorentey commented:
There were no issues left in my local validation test runs -- I ran them on Linux/x86_64, macOS/x86_64 and watchOS-simulator/i386 for [#18](https://github.com/milseman/swift/pull/18).

I just finished a compiler build on Linux from the latest sources. Subscripting with out-of-bounds indices correctly triggers the precondition on [StringVariant.swift:90](https://github.com/milseman/swift/blob/string_guts/stdlib/public/core/StringVariant.swift#L90) and `build-script -rT` reports no issues (after removing a hang on a compiler crasher test). I'm looking into what's different in the buildbot_linux environment.

milseman commented:
@swift-ci please test Linux Platform

lorentey commented:
@swift-ci please test Linux Platform

lorentey committed:
String & String views: Add bounds checking to range subscripts. (#19)

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
Err, MacOS was just passing.

@swift-ci please test os x platform.

milseman commented:
This all passes locally for me with ./utils/build-script -R -T -i. I'll also try a --no-assertions...

@swift-ci please test OS X platform

milseman commented:
@swift-ci please test os x platform

milseman committed:
WIP: work around compiler bug by using @_fixed_layout

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
Failures:
```
    Swift(iphonesimulator-i386) :: stdlib/AnyHashable.swift.gyb
    Swift(iphonesimulator-i386) :: Interpreter/SDK/UIActionSheetHack.swift
    Swift(iphonesimulator-i386) :: stdlib/CodableTests.swift
    Swift(iphonesimulator-i386) :: stdlib/TestJSONEncoder.swift
```

milseman commented:
So I built locally identically to the bot, and these don't consistently reproduce.

@swift-ci please clean test os x platform

milseman commented:
Rebased, re-built locally, still cannot repro...

@swift-ci please test os x platform

lorentey commented on stdlib/public/SDK/Foundation/NSString.swift:
```diff
@@ -23,8 +23,8 @@ public class NSSimpleCString {}
-public func _convertStringToNSString(_ string: String) -> NSString {
-  return string._bridgeToObjectiveC()
+public func _getDescription<T>(_ x: T) -> NSString {
+  return String(reflecting: x)._bridgeToObjectiveC()
```
@milseman Oops, sorry, I missed this question earlier. This change merges `_getSummary` in the stdlib with `_convertStringToNSString` in Foundation to eliminate the need to hold a `String` value in the runtime.

`_getSummary` itself used `String(reflecting:)`, I just moved it over here.

milseman committed:
[Builtin] Add _isValidAddress pointer-value check.

And adopt it for StringGuts, as it is more robust on 32-bit systems.

milseman commented:
(Tiny bit of a shot in the dark, but I suspect it's 32-bit pointers in the latter half of the address space).

@swift-ci please test

milseman commented:
Got past iphonesimulator-i386!
🎉

milseman commented:
Woo! Going to push some squashes.

milseman commented:
Merge will take place in https://github.com/apple/swift/pull/14046
